New manpage format

jnavila opened the pull request:
## Changes

This PR changes the way the asciidoc source of manpage is processed, by adding the "synopsis" paragraph style and reworking the backtick format.

## Context

The style change has been pushed to master and will be applied to git-clone and git-init in the next version.

jnavila committed:
manpages: prepare for new manpage format

This commit adds a upcoming manpage format to the AsciiDoc
backend. The new format changes are:

 * The synopsis is now a section with a dedicated style. This
 "synopsis" style allows to automatically format the keywords as
 monospaced and <placeholders> as italic.
 * the backticks are now used to format synopsis-like syntax in inline
 elements.

The parsing of synopsis is done with a new AsciiDoc extension that
makes use of the PEG parser parslet.

All the asciidoc manpages sources are processed with this
extension. It may upset the formatting for older manpages, making it
not consistent across a page, but this will be a mild side effect, as
this was not really consistent before.

Signed-off-by: Jean-Noël Avila <jn.avila@free.fr>

dscho commented:
I just triggered a pair of workflow runs to [update the manual pages](https://github.com/dscho/git-scm.com/actions/runs/11898105983/job/33153838707) and to [update the translated manual pages](https://github.com/dscho/git-scm.com/actions/runs/11898466848), fetched the result and rendered it locally. Here are two examples:

| language | before | after|
| - | - | - |
| English | ![image](https://github.com/user-attachments/assets/b4bb11f3-0b5c-415b-92d3-574fbbdb624e) | ![image](https://github.com/user-attachments/assets/3a0e19b5-892c-4710-94e8-52cc7d26184c) |
| French | ![image](https://github.com/user-attachments/assets/e8280585-29ff-43c7-8849-68e1b4eb9fbf) | ![image](https://github.com/user-attachments/assets/088b34d1-1497-4f61-ada3-69dcd160d053) |

Personally, I cannot spot any difference, apart from the version number (because this here PR branch is based on v2.46.2 while the updated manual pages include v2.47.0) and the incorrect `=<regexp>` on the "before" side of the French version (fixed on the "after" side).

Even looking at the HTML of the synopses (taking the French version, so that there _is_ a known difference), I only see this:

```diff
diff --git a/before b/after
index 1a87d1348..6185fb72b 100644
--- a/before
+++ b/after
@@ -1,5 +1,5 @@
 <pre class="content"><em>git config list</em> [&lt;option-de-fichier&gt;] [&lt;option-d-affichage&gt;] [--includes]
-<em>git config get</em> [&lt;option-de-fichier&gt;] [&lt;option-d-affichage&gt;] [--includes] [--all] [--regexp=&lt;regexp&gt;] [--value=&lt;valeur&gt;] [--fixed-value] [--default=&lt;default&gt;] &lt;nom&gt;
+<em>git config get</em> [&lt;option-de-fichier&gt;] [&lt;option-d-affichage&gt;] [--includes] [--all] [--regexp] [--value=&lt;valeur&gt;] [--fixed-value] [--default=&lt;default&gt;] &lt;nom&gt;
 <em>git config set</em> [&lt;option-de-fichier&gt;] [--type=&lt;type&gt;] [--all] [--value=&lt;valeur&gt;] [--fixed-value] &lt;nom&gt; &lt;valeur&gt;
 <em>git config unset</em> [&lt;option-de-fichier&gt;] [--all] [--value=&lt;valeur&gt;] [--fixed-value] &lt;nom&gt; &lt;valeur&gt;
 <em>git config rename-section</em> [&lt;option-de-fichier&gt;] &lt;ancien-name&gt; &lt;nouveau-name&gt;
```

@jnavila what am I missing?

jnavila commented:
The manpage of git-config has not been converted yet. 
I pushed a branch "test-refactor" on git-html-l10n, where I hand-edited `fr/git-add.txt`.

After importing, here is the result:

![image](https://github.com/user-attachments/assets/586fce8e-065f-4c8e-8ae1-96b81651f48a)

I'm not satisfied with the styles, particularly when dealing with inline formats:

![image](https://github.com/user-attachments/assets/3b14f527-2875-428a-915a-250ccb26c0df)


you can test by yourself locally, and tell me your judgment.

jnavila committed:
stylesheet: remove background and border from code spans

The new style makes the code spans lighter and more integrated into
the text and also makes the code spans more readable and
less intrusive.

A synopsis class is added for the code spans that are synopsis.

Signed-off-by: Jean-Noël Avila <jn.avila@free.fr>

jnavila commented:
@dscho I updated the CSS, so it is ready for review.

To1ne commented on script/asciidoctor-extensions.rb:
```diff
@@ -0,0 +1,63 @@
+
+      def process parent, reader, attrs
+        outlines = reader.lines.map do |l|
+          l.gsub(/(\.\.\.?)([^\]$.])/, '`\1`\2')
```
I think a line of comment wouldn't hurt with these regexes. Maybe best with an example:

```suggestion
          l.gsub(/(\.\.\.?)([^\]$.])/, '`\1`\2') # wrap ellipsis in backticks: ...something => `...`something
```

I think the intended use is for `[...<more>]`? Should we include the `[` and `]` in the regex?

jnavila replied:
This line is trying to differentiate the three dots in different contexts, where they have a different meaning and require different formatting.

First there is the form \<commit1\>...\<commit2\> when describing a range of commits, where the three dots are a "keyword" understood by git and must be formatted as code.

Then there is the forms used in the grammar to express repetition, such as in "\<path\> ..." with optionally square brackets, such as "[\<path\>...]" which usually appear at the end of the command line. These three dots must not be formatted as code, but left as is.

This line matches the former case and forces the corresponding format. I'll add  a comment in the same line as yours.

To1ne commented on script/asciidoctor-extensions.rb:
```diff
@@ -0,0 +1,63 @@
+      def process parent, reader, attrs
+        outlines = reader.lines.map do |l|
+          l.gsub(/(\.\.\.?)([^\]$.])/, '`\1`\2')
+           .gsub(%r{([\[\] |()>]|^)([-a-zA-Z0-9:+=~@,/_^\$]+)}, '\1{empty}`\2`{empty}')
```
To be honest, I don't know what this one is for.

jnavila replied:
This one is the line that matches all the words which are not placeholders and not grammar signs, and format them as code. These words (in the general sense here) are keywords (option names, enum strings, two or three dot notation, etc).

To1ne commented on script/asciidoctor-extensions.rb:
```diff
@@ -0,0 +1,63 @@
+        outlines = reader.lines.map do |l|
+          l.gsub(/(\.\.\.?)([^\]$.])/, '`\1`\2')
+           .gsub(%r{([\[\] |()>]|^)([-a-zA-Z0-9:+=~@,/_^\$]+)}, '\1{empty}`\2`{empty}')
+           .gsub(/(<([[:word:]]|[-0-9.])+>)/, '__\\1__')
```
I had to dig deep to find what `[[:word:]]` does, but it seems to be a Ruby non-POSIX bracket expression: https://docs.ruby-lang.org/en/master/Regexp.html#class-Regexp-label-POSIX+Bracket+Expressions. Personally I'm not a fan, what's the advantage over `\w`?

Also why are the inner brackets round brackets? 

I just wonder if we can simplify to:

```suggestion
           .gsub(/(<[^>]+>)/, '__\\1__')
```

And one more question, why the double backslash in the replacement string?

jnavila replied:
The '\w` is for ascii, but here, we are going to process internationalized texts (because placeholders are translated), and this processing requires the special form with double brackets. I'm not an expert in Ruby regexes; this is the form I have found to work well with the translations.

As for the using a more generic regex (expecting everything between brackets to be the placeholder's name), the placeholder's names are not supposed to contain spaces, which is perfect when we have to match something like:
```
 $ git foo < in-file > out-file
```

To1ne commented on script/asciidoctor-extensions.rb:
```diff
@@ -0,0 +1,63 @@
+        if node.type == :monospaced
+          node.text.gsub(/(\.\.\.?)([^\]$.])/, '<code>\1</code>\2')
+              .gsub(%r{([\[\s|()>.]|^|\]|&gt;)(\.?([-a-zA-Z0-9:+=~@,/_^\$]+\.{0,2})+)}, '\1<code>\2</code>')
+              .gsub(/(&lt;([[:word:]]|[-0-9.])+&gt;)/, '<em>\1</em>')
```
So we more or less need to repeat the regexes here?

jnavila replied:
That's unfortunate, but the two regex are very alike, except that this one processes the text after some pre-processing steps, and the transformations need to be in final result form (with tags and escaped characters).

I evaluated the opportunity for factorization, but it makes the code more messy than it is already.

To1ne reviewed (commented):
@jnavila I've added a few questions. Thanks for this contribution.

dscho reviewed (commented):
I am really uneasy with this large amount of hard-to-understand regular expressions. Not only makes this bugs easy to hide, it also inadvertently opens the door to DoS attacks. [Here](https://stackstatus.tumblr.com/post/147710624694/outage-postmortem-july-20-2016) is an example where something like this has had a really high impact.

It would probably make much more sense to implement a `StringScanner`-based parser that is much easier to reason about and whose performance is well-understood, e.g. following [this tutorial](https://www.rubyguides.com/2015/04/parsing-with-ruby/).

jnavila commented:
These regexes are basically the same ones that I already pushed to git/git. They are applied during the conversion phase, not in live, and only on the quoted strings of text, after the initial asciidoc parsing has been performed.

 Anyway, as they are cryptic, I can try to convert the code to a parser, but I doubt this will be a lot clearer.

jnavila commented:
Writing a parser turns out to be more involving than expected, mainly because the new parser must be resilient to formatting mistakes in the translations (particularly in Chinese) and manpages that were not converted to the new format (where the backticks are used for widely varying strings).

dscho commented:
A parser would at least document a lot more cohesively what is being done.

As they are, I know what the regular expressions do, but only after studying them extensively, an effort that I would most likely have to repeat were I in the need to fix any bugs in that code in the future. I am extremely uncomfortable with that, source code should be as obvious as possible, and this is not it. As such, I will not review this any further and not merge it. I won't oppose anybody else merging it, but I want nothing to do with the added code, myself.

jnavila commented:
I've pushed an updated version with a parser. In case the parser fails, we just roll back to the basis code format.

jnavila commented:
@dscho Is this version better for you?

dscho commented:
@jnavila I am sorry, I had not planned any involvement in this anymore, due to time constraints.

To1ne commented:
Ah cool, I just realize you're trying to fix https://github.com/git/git-scm.com/issues/1972 here.

To1ne commented:
@jnavila I've noticed we have a synopsis processor at https://github.com/git/git/blob/master/Documentation/asciidoctor-extensions.rb.in. Why did you chose to write it yourself? I rather not maintain 2 versions.

I've used that extension locally (without the postprocessor) and this is how the synopsis on git-clone looks:

![image](https://github.com/user-attachments/assets/8ae13398-ebce-48a0-8afc-51d1b245b419)

Or on git-diff-index:

![image](https://github.com/user-attachments/assets/58d12d4b-6471-4438-8271-f1bd3d7f8f95)

I feel it's very much a "drop-in" extension we can/should use.

jnavila commented:
> @jnavila I've noticed we have a synopsis processor at https://github.com/git/git/blob/master/Documentation/asciidoctor-extensions.rb.in. Why did you chose to write it yourself? I rather not maintain 2 versions.

In fact, I wrote both of the extensions.. :laughing:  . The rewrite here is a request from @dscho which is very convincing to me. The synopsis is a language in itself, so it is more comprehensive to have a dedicated parser for this language instead of a bunch of repelling regexps.

This rewrite came after I proposed the regexp version for inclusion in git/git . Now, I'm contemplating reverting the git/git version to the parser flavor, and at the same time removing this ugly pre-processor hack by a38edab.

dscho commented on script/asciidoctor-extensions.rb:
```diff
@@ -0,0 +1,129 @@
+      rule(:opt_or_alt) { match('[\[\] |()]') >> space? }
+      rule(:ellipsis) { str('...') >> match('\]|$').present? }
+      rule(:grammar) { opt_or_alt | ellipsis }
+      rule(:ignore) { match('[\'`]') }
```
I have to admit that I was quite thrown by the syntax, especially the `>>` one. https://kschiess.github.io/parslet/parser.html to the rescue (which we might want to link to, in a code comment).

Narrators voice: The `>>` indicates a "simple sequence", for example `str('...') >> match('\]|$').present?` means "first match three periods, then ensure that they are either followed by a closing bracket or they are at the end.

dscho commented on script/asciidoctor-extensions.rb:
```diff
@@ -0,0 +1,129 @@
+      rule(:ignore) { match('[\'`]') }
+
+      rule(:token) do
+        grammar.as(:grammar) | placeholder.as(:placeholder) | space.as(:grammar) |
```
Shouldn't this be `space.as(:space)`?

jnavila replied:
grammar and space are left unchanged, so I put them in the same bag.

dscho replied:
You mean `grammar` is handled by [this line](https://github.com/git/git-scm.com/pull/1921/files/d9070e669020ab7fd0c8dc4c24cc42a33545e135..87aab82c6e263f15631de2423db0ad6a55783338#diff-520e8c3aafb0fe1371224cd2062b800117736dd6a9f0dd2b3c7008c50634ee8aR47) and [this line](https://github.com/git/git-scm.com/pull/1921/files/d9070e669020ab7fd0c8dc4c24cc42a33545e135..87aab82c6e263f15631de2423db0ad6a55783338#diff-520e8c3aafb0fe1371224cd2062b800117736dd6a9f0dd2b3c7008c50634ee8aR54) by leaving the original text unchanged?

It still might be clearer to introduce a corresponding `rule(space: simple(:space)) { space.to_s }` line (and to let `SynopsisQuoteToHtml5 ` inherit from `SynopsisQuoteToAdoc`, overriding only `keyword` and `placeholder`).

dscho commented on script/asciidoctor-extensions.rb:
```diff
@@ -0,0 +1,129 @@
+      rule(:space)      { match('[\s\t\n ]').repeat(1) }
+      rule(:space?)     { space.maybe }
+      rule(:keyword) { match('[-a-zA-Z0-9:+=~@,\./_\^\$\'"\*%!{}#]').repeat(1) }
+      rule(:placeholder) { str('&lt;') >> match('[[:word:]]|-').repeat(1) >> str('&gt;') }
```
If this is the only difference to `AdocSynopsisQuote`, why not use `class EscapedSynopsisQuote <AdocSynopsisQuote` and override just this rule?

jnavila replied:
Ah, I haven't tried to use inherit between my classes.

dscho replied:
I hope it will work, otherwise I'd have caused you a lot of effort for nothing in return.

dscho commented on script/asciidoctor-extensions.rb:
```diff
@@ -0,0 +1,129 @@
+      rule(:ignore) { match('[\'`]') }
+
+      rule(:token) do
+        grammar.as(:grammar) | placeholder.as(:placeholder) | space.as(:grammar) |
```
Again, should this be `space.as(:space)`?

dscho reviewed (commented):
Wow. This sure is a lot more verbose, but _vastly_ easier to understand than the regular expression-based solution. Thank you for putting in the work.

I have a couple of questions/suggestions below, and also the request: Please rebase to current `gh-pages` (resolving the merge conflict in `Gemfile`, too). Thank you!

To1ne commented:
> In fact, I wrote both of the extensions.. 😆 . The rewrite here is a request from @dscho which is very convincing to me.  The synopsis is a language in itself, so it is more comprehensive to have a dedicated parser for this language instead of a bunch of repelling regexps.

@jnavila I'm sorry for dropping in completely ignorant!

> This rewrite came after I proposed the regexp version for inclusion in git/git . Now, I'm contemplating reverting the git/git version to the parser flavor

I haven't looked at it in detail yet, but at first sight that makes sense.

> and at the same time removing this ugly pre-processor hack by a38edab.

Do you mean https://github.com/git/git/commit/a38edab7c88b5503bb2b5f5cbd49f6b97e9a6a4e ? If you mean the hack to fill in  `@GIT_VERSION@`, I agree it would be nice to get rid of it. Although it's probably not directly related to the implementation details of the parser.

That said, can we deduplicate having a synopsis parser here and one in git/git.git ?

To1ne commented:
> After importing, here is the result:
> ...
>I'm not satisfied with the styles

@jnavila Without thinking about how to achieve this, how do you prefer the end result will look like? In a quick mockup I've ended up with the following:

![image](https://github.com/user-attachments/assets/8327ebf2-5299-4034-b48f-ecc4ad2cb050)

Is that something you'd like?

jnavila commented:
That would be great 😃. The parser cannot make a difference between the "git commit" part and the options, so they would have the same style ( bold red… for instance)

dscho commented:
> The parser cannot make a difference between the "git commit" part and the options

Couldn't the parser learn that Git commands start with `git` and then continue with white-space followed by a command-name that matches `[a-z][-0-9a-z]*` (yes, `git p4` contains digits)? That should make it possible to discern between commands and options quite nicely.

To1ne commented:
@jnavila I see there's more work being done in this direction: https://lore.kernel.org/git/20250501213414.370514-1-gitster@pobox.com/. I'm sorry for my lazy question, but what is the status of this PR? What do you/we need to drive it forward?

jnavila commented:
@To1ne In its current state, I think the PR is already able to correctly parse and format all the elements in back-quotes. There was a request to be able to format more specifically the "git \<command\>" part, but I didn't find time to work on it.

We may merge this PR and open another issue with the requested enhancement.

jnavila commented:
I should rebase and address the comments from @dscho though.

To1ne commented:
> In its current state, I think the PR is already able to correctly parse and format all the elements in back-quotes. 

For your entertainment deployed to https://to1ne.github.io/git-scm.com/docs/git-clone

jnavila commented:
> That said, can we deduplicate having a synopsis parser here and one in git/git.git ?

This can be pushed also to git/git, but only for asciidoctor. For asciidoc, we are stuck with regex.

dscho commented:
@jnavila @To1ne what's the status on this? Do we need a fix for #2002 before merging this (preemptively addressing the `<placeholder>` tag)?

jnavila commented:
The PR was reworked according to your remarks. I won't have much time to work on adding the code to deal with "git foo" formatting. Another span tag is added, so #2002 really needs to be fixed before.

dscho commented:
> Another span tag is added, so #2002 really needs to be fixed before.

@To1ne do you think you'd have time to work on that? I won't have time for at least another week... 😦

dscho commented:
@jnavila you probably want to rebase to `gh-pages`, force-push and then run [the `update-git-version-and-manual-pages` workflow](https://github.com/git/git-scm.com/actions/workflows/update-git-version-and-manual-pages.yml) and then [the `update-translated-manual-pages` workflow](https://github.com/git/git-scm.com/actions/workflows/update-translated-manual-pages.yml), both times forcing a full rebuild by checking that checkbox.

dscho commented:
@jnavila FYI I canceled [the `gh-pages` run](https://github.com/git/git-scm.com/actions/runs/16073267146) (because it would not have picked up the changes from this here PR) and triggered a [new, `new_manpage_format` run](https://github.com/git/git-scm.com/actions/runs/16073318219).

dscho commented:
Ouch.

dscho commented:
Now the run I triggered will be in vain.

jnavila commented:
OK Obviously, I didn't understand what your comment meant.
What was I suppose to do?

dscho commented:
@jnavila FYI I canceled the run that would have updated the now-deleted PR branch and triggered a new `gh-pages` one.

Next time, let's do let the workflow make the updates _before_ the PR is merged. That way, we can verify (and validate) the results before integrating them into the default branch.

jnavila commented:
Ah, ok. I wasn't allowed to run the workflow on the PR branch, so I thought (wrongly) it was only possible on `gh-pages`.

dscho commented:
> I wasn't allowed to run the workflow on the PR branch

@jnavila it is true that by default, the `gh-pages` environment (which is required to deploy) is limited to the `gh-pages` branch. I had temporarily added `new_manpage_format` to that environment to allow it to run. It's now removed again, of course.

jnavila commented:
Anyway, this PR requires a follow-up because the tokens such as `%(token)` are not processed correctly. I'm trying to come up with something understandable in the upstream git/git repo with regexes, and it is difficult. Comparatively, I expect it to be piece of cake with the parser.
